store/tikv/latch/latch.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package latch
+
+import (
+	"fmt"
+	"math"
+	"sort"
+	"sync"
+
+	"github.com/spaolacci/murmur3"
+)
+
+// Latch stores a key's waiting transactions.
+type Latch struct {
+	// A queue by startTs of those waiting transactions.
+	waiting      []uint64
+	lastCommitTs uint64
+	sync.Mutex
+}
+
+// acquire tries to get current key's lock for the transaction with startTS.
+// acquire is true when success
+// timeout is true when the startTS is already timeout
+// newWait is true when current transaction is new for the current latch
+func (l *Latch) acquire(startTS uint64) (acquire, timeout, newWait bool) {
+	l.Lock()
+	defer l.Unlock()
+	timeout = startTS <= l.lastCommitTs
+	if timeout {
+		return
+	}
+	if len(l.waiting) == 0 || l.waiting[0] != startTS {
+		l.waiting = append(l.waiting, startTS)
+		newWait = true
+	}
+
+	acquire = l.waiting[0] == startTS
+	return
+}
+
+// release releases the transaction with startTS and commitTS from current latch.
+// isEmpty is true when the waiting queue is empty after release current transaction,
+// otherwise return the front transaction in queue.
+func (l *Latch) release(startTS uint64, commitTS uint64) (isEmpty bool, front uint64) {
+	l.Lock()
+	defer l.Unlock()
+	if startTS != l.waiting[0] {
+		panic(fmt.Sprintf("invalid front ts %d, latch:%+v", startTS, l))
+	}
+	if commitTS > l.lastCommitTs {
+		l.lastCommitTs = commitTS
+	}
+	l.waiting = l.waiting[1:]
+	if len(l.waiting) == 0 {
+		isEmpty = true
```

Comment:
I guess better to reset the slice by `l.waiting = l.waiting[:0]`? @tiancaiamao

Reply:
if `l.waiting == nil`, `l.waiting = l.waiting[:0]` would panic.

Reply:
if `l.waiting == nil`, it will panic at L65.